pkg/security/controller/namespace_scc_allocation_controller.go

```diff
@@ -0,0 +1,339 @@
+		return nil
+	}
+
+	return c.allocate(ns)
+}
+
+func (c *NamespaceSCCAllocationController) allocate(ns *corev1.Namespace) error {
+	// unless we affirmatively succeed, clear the local state to try again
+	success := false
+	defer func() {
+		if success {
+			return
+		}
+		c.currentUIDRangeAllocation = nil
+	}()
+
+	// if we don't have the current state, go get it
+	if c.currentUIDRangeAllocation == nil {
+		newRange, err := c.rangeAllocationClient.RangeAllocations().Get(rangeName, metav1.GetOptions{})
+		if err != nil {
+			return err
+		}
+		c.currentUIDRangeAllocation = newRange
+	}
+
+	// do uid allocation.  We reserve the UID we want first, lock it in etcd, then update the namespace.
+	// We allocate by reading in a giant bit int bitmap (one bit per offset location), finding the next step,
+	// then calculating the offset location
+	uidRange, err := uid.ParseRange(c.currentUIDRangeAllocation.Range)
+	if err != nil {
+		return err
+	}
+	if !reflect.DeepEqual(*uidRange, *c.requiredUIDRange) {
+		return fmt.Errorf("conflicting UID range; expected %#v, got %#v", *c.requiredUIDRange, *uidRange)
+	}
+	allocatedBitMapInt := big.NewInt(0).SetBytes(c.currentUIDRangeAllocation.Data)
+	bitIndex, found := allocateNextContiguousBit(allocatedBitMapInt, int(uidRange.Size()))
+	if !found {
+		return fmt.Errorf("uid range exceeded")
+	}
+	allocatedBitMapInt = allocatedBitMapInt.SetBit(allocatedBitMapInt, bitIndex, 1)
+	newRangeAllocation := c.currentUIDRangeAllocation.DeepCopy()
+	newRangeAllocation.Data = allocatedBitMapInt.Bytes()
+
+	actualRangeAllocation, err := c.rangeAllocationClient.RangeAllocations().Update(newRangeAllocation)
+	if err != nil {
+		return err
+	}
+	c.currentUIDRangeAllocation = actualRangeAllocation
+
+	block, ok := uidRange.BlockAt(uint32(bitIndex))
+	if !ok {
+		return fmt.Errorf("%d not in range", bitIndex)
+	}
+
+	// Now modify the namespace
+	nsCopy := ns.DeepCopy()
+	if nsCopy.Annotations == nil {
+		nsCopy.Annotations = make(map[string]string)
+	}
+	nsCopy.Annotations[security.UIDRangeAnnotation] = block.String()
+	nsCopy.Annotations[security.SupplementalGroupsAnnotation] = block.String()
+	if _, ok := nsCopy.Annotations[security.MCSAnnotation]; !ok {
+		if label := c.mcsAllocator(block); label != nil {
+			nsCopy.Annotations[security.MCSAnnotation] = label.String()
+		}
+	}
+
+	_, err = c.namespaceClient.Update(nsCopy)
+	if apierrors.IsNotFound(err) {
+		return nil
+	}
+	if err != nil {
+		return err
+	}
+	success = true
+	return nil
+}
+
+// allocateNextContiguousBit finds a free bit in the int and returns which one it is and whether it succeeded
+func allocateNextContiguousBit(allocated *big.Int, max int) (int, bool) {
+	for i := 0; i < max; i++ {
+		if allocated.Bit(i) == 0 {
+			return i, true
+		}
+	}
+	return 0, false
+}
+
+func (c *NamespaceSCCAllocationController) WaitForRepair(stopCh <-chan struct{}) error {
+	return wait.PollImmediate(10*time.Second, 5*time.Minute, func() (bool, error) {
```

Comment:
> can you explain these intervals? Why we need to wait and not just reconcile?

It's arbitrary, but you need proper current state before you start running the controller overall.  This is the same as the previous.   You need to make sure the list of allocated UIDs is accurate before you start assigning more.



> in case of uidallocator.ErrFull this will just loop for 5 minutes?

And then fail.  Yes.  It's a very uncommon edge.  We have something like 100k slots